src/InputSource/index.ts

```diff
@@ -0,0 +1,50 @@
+import fs from "fs";
+import reader from "readline-sync";
+
+import { Optional } from "../util/types";
+
+export type InputSourceType = "file" | "cli";
+
+export interface InputSource {
+  getInput(): Optional<string>;
+}
+
+export class InputSourceFactory {
+  static createFileInputSource(source: string): FileInputSource {
+    return new FileInputSource(source);
+  }
+
+  static createCliInputSource(): InputSource {
+    return new CliInputSource();
+  }
+}
+
+class FileInputSource implements InputSource {
+  private source: string;
+
+  constructor(source: string) {
+    this.source = source;
+  }
+
+  getInput(): Optional<string> {
```

Comment:
I don't think you need to specify return type for `getInput`implementations, most likely TS will infer them automatically